### DISCUSSION TITLE
Trying to add vertical bar swinging animation, worried it's going to open a can of worms.

### DISCUSSION THREAD

**Discussion by DimensionWarped** [Techokami/SonicWorldsNext#11] [Q&A] [answered]
I was trying to add the vertical bar swinging animation used a few times throughout Sonic and Knuckles (specifically, there is a simple vertical bar in Mushroom Hill Zone that you swing on once before launching forward as well as these large propeller things in Flying Battery that are similar, but much thicker... those details aren't super important though).

I noticed that the existing sprites already have frames for these animations, but unlike most of the frames in the animation, they aren't always offset properly. I'm still extremely new to Godot, so maybe I'm missing something obvious (and that's what I'm here hoping to find out), but if I'm not is it the case that the entire sprite sheet will need to be expanded in order to widen the space that frames can take up so that these animations can be centered at the hands?

![image](https://user-images.githubusercontent.com/7477617/214027105-a4563350-b923-4f66-80ef-ea8a28ca3a47.png)
![image](https://user-images.githubusercontent.com/7477617/214027134-f34d4430-404e-4e4a-8260-0f2e8582d4eb.png)

Attached are a couple pictures showing the frames in question.

**Comment by Renhoex** [comment, marked as answer]
It is possible to center the animations, you just gotta keyframe the offsets.
Tails actually does this for his offset corkscrew animation to stay centered, just make sure offset is set to discrete in the animator and that the offset is reset when the animation's finished or when the state gets changed back.

it might help to look through the corkscrew gimmick for how to get that last part done right.
![Capture](https://user-images.githubusercontent.com/26295900/214030898-b2e18b52-91b9-40fb-af90-85202a5107c4.PNG)

Hope this helps.

**Comment by DimensionWarped** [comment]
Thanks. I'll review Tails animations around the corkscrew and try to figure out how that works and get back to you guys if I I'm still struggling to make it work.

**Reply by DimensionWarped** [reply]
@Renhoex 

Can you confirm for me real quick if this is right?


```
diff --git a/Source/Graphics/Players/PlayerAnimations/Sonic.tscn b/Source/Graphics/Players/PlayerAnimations/Sonic.tscn
index ee7d2a6..f90ca56 100644
--- a/Source/Graphics/Players/PlayerAnimations/Sonic.tscn
+++ b/Source/Graphics/Players/PlayerAnimations/Sonic.tscn
@@ -1,4 +1,4 @@
-[gd_scene load_steps=47 format=2]
+[gd_scene load_steps=48 format=2]

 [ext_resource path="res://Shaders/PlayerPalette.tres" type="Material" id=1]
 [ext_resource path="res://Graphics/Players/Sonic.png" type="Texture" id=2]
@@ -577,6 +577,34 @@ tracks/1/keys = {
 "values": [ 1.0, 1.0, 9.0 ]
 }

+[sub_resource type="Animation" id=121]
+resource_name = "swing_vertical_bar"
+loop = true
+tracks/0/type = "value"
+tracks/0/path = NodePath("Sprite/Sprite:frame")
+tracks/0/interp = 1
+tracks/0/loop_wrap = true
+tracks/0/imported = false
+tracks/0/enabled = true
+tracks/0/keys = {
+"times": PoolRealArray( 0, 0.2, 0.4, 0.6, 0.8, 1 ),
+"transitions": PoolRealArray( 1, 1, 1, 1, 1, 1 ),
+"update": 1,
+"values": [ 213, 208, 209, 210, 211, 212 ]
+}
+tracks/1/type = "value"
+tracks/1/path = NodePath("Sprite/Sprite:offset")
+tracks/1/interp = 1
+tracks/1/loop_wrap = true
+tracks/1/imported = false
+tracks/1/enabled = true
+tracks/1/keys = {
+"times": PoolRealArray( 0, 0.2, 0.4, 0.6, 0.8, 1 ),
+"transitions": PoolRealArray( 0.784584, 1, 1, 1, 1, 1 ),
+"update": 1,
+"values": [ Vector2( 0, -4 ), Vector2( 20, -4 ), Vector2( 20, -4 ), Vector2( -20, -4 ), Vector2( -20, -4 ), Vector2( -20, -4 ) ]
+}
+
 [sub_resource type="Animation" id=85]
 resource_name = "walk"
 length = 0.8
@@ -671,6 +699,7 @@ anims/spinDash = SubResource( 83 )
 anims/spring = SubResource( 84 )
 anims/springScrew = SubResource( 119 )
 anims/super = SubResource( 97 )
+anims/swing_vertical_bar = SubResource( 121 )
 anims/walk = SubResource( 85 )

 [node name="Sprite" type="Sprite" parent="."]
```

note: this is Sonic only, but I'll be creating a pull request once I have all characters finished.

**Reply by Renhoex** [reply]
I can't really gauge this but I think it's ok, if it all works then it should be fine.

**Comment by DimensionWarped** [comment]
One other question, is there a place where we need to adjust hurt boxes to make them synch up with animations?

**Reply by Renhoex** [reply]
In the player script there's a list of hitbox sizes at the start of the script for each character.
under the "set_state" function there's a switch for setting the masks sizes depending on the state it's swapping to

**Reply by Renhoex** [reply]
We have plans for doing a document, I'll try to remember to cover this when we get around to it.

**Reply by DimensionWarped** [reply]
Is it possible to set that per frame or only per state?

EDIT: Actually what I think this particular animation is going to need is less continually setting sizes and more continually setting positions... though I guess both *should* change for this animation.

**Reply by Renhoex** [reply]
It's currently per state but it would be possible to code it in per frame if you coded it into the state script.

I don't know how familiar you are with the games mechanics but if you need a reference for how the hitboxes work there's this overlay script that can show you how sonic's hitbox works.

http://info.sonicretro.org/SPG:Overlay_Scripts

just a heads up, replies might be slow from here I might not be able to respond for a while

**Comment by DimensionWarped** [comment]
@Renhoex 

Hey, sorry to bother, but I've got a rather odd situation that I'm having trouble isolating on my attempt to implement the vertical swinging bar. I'm about 80% done with it... the main thing remaining is the dismount, but my issue is that for some crazy reason Tails offset is wrong, but only if he is player 2. If Tails is chosen as the player character then it works normally.

You can look at my fork here...
https://github.com/DimensionWarped/SonicWorldsNext

Nothing I've done changes the player 2 stuff as far as I can tell and I'm totally stumped on where to even start looking.

**Comment by Renhoex** [comment]
This took me a minute to figure out but I figured it out, the players movement isn't changed when they're running into the pole, since player 2 is added to the end of the node tree his code is executed after the pole so tails ends up continuing with whatever momentum he had after he goes into the pole animation.

to fix it add `i.movement.x = 0` into the code when the player grabs the pole.

Looking through the code so far it's actually pretty clean for the most part, nice work so far keep it up!

**Reply by DimensionWarped** [reply]
Ah, that worked. Thanks!
